src/GovUk.Education.ExploreEducationStatistics.Data.Api/Controllers/PermalinkController.cs

```diff
@@ -33,7 +34,8 @@ public PermalinkController(IPermalinkService permalinkService)
                     contentType: "text/csv",
                     filename: $"permalink-{id}.csv");
 
-                await _permalinkService.LegacyDownloadCsvToStream(id, Response.BodyWriter.AsStream(), cancellationToken);
+                await _permalinkService.LegacyDownloadCsvToStream(id, Response.BodyWriter.AsStream(),
+                    cancellationToken);
```

Comment:
I'm just wondering if you tried this code, (which I'm guessing you did as you even included the unit tests for it!), as I couldn't get it to work.

There seems to be two problems with the suggested code.

1.  As soon as the stream is touched by `LegacyDownloadCsvToStream` the response starts being written. `ExecuteResultAsync` will then fail with error `The status code cannot be set, the response has already started`.
2. The result of `HandleFailuresOr(Ok)` is an `OkObjectResult` with value `Unit`. That's not compatible with the client's Accept header `text/csv` , so if `ExecuteResultAsync` could complete it would produce a 406 Not-Acceptable result instead.

The only way that I could get this to work was by replacing

```
            if (Request.AcceptsCsv(exact: true))
            {
                Response.ContentDispositionAttachment(
                    contentType: "text/csv",
                    filename: $"permalink-{id}.csv");

                await _permalinkService.LegacyDownloadCsvToStream(id, Response.BodyWriter.AsStream(),
                    cancellationToken);

                return;
            }
```

with

```
            if (Request.AcceptsCsv(exact: true))
            {
                using var stream = new MemoryStream();
                var csvResult = await _permalinkService.LegacyDownloadCsvToStream(
                        id,
                        stream,
                        cancellationToken
                    )
                    .HandleFailuresOr(_ =>
                    {
                        stream.Position = 0;
                        return File(stream, "text/csv", $"permalink-{id}.csv");
                    });

                await csvResult.ExecuteResultAsync(ControllerContext);

                return;
            }
```

where `File` is `ControllerBase.File` which returns a `FileStreamResult`.

Reply:
Converting this to a `FileStreamResult` isn't a good idea as it basically means we have to buffer the entire permalink into memory rather than streaming it directly to the response.

I think this should work however:

```cs
Response.ContentDispositionAttachment(
    contentType: "text/csv",
    filename: $"permalink-{id}.csv");

var csvResult = await _permalinkService.LegacyDownloadCsvToStream(
    permalinkId: id,
    stream: Response.BodyWriter.AsStream(),
    cancellationToken: cancellationToken
)
    .HandleFailuresOr(Ok);

if (csvResult is not OkObjectResult)
{
    await csvResult.ExecuteResultAsync(ControllerContext);
}
```

Similar thing could be done for the newer permalink endpoint.

Reply:
That does appear to work, thanks for the update and the demo yesterday. 🙂 Made the change in both places.